### PR TITLE
FEM-2357 Add External (out-of-band) WebVTT captions support

### PR DESCRIPTION
**Pull request by x-NR-x**
Supports external subtitles for media which has no internal subtitles.

- Moved AssetHandler and DefaultAssetHandler under the same folder named AVAsset.
- Added a PKAssetResourceLoaderDelegate to hold all optional delegates.
- Updated the DefaultAssetHandler to have a PKAssetResourceLoaderDelegate.
The avAsset will always have the PKAssetResourceLoaderDelegate when set.
Code was updated accordingly for the FP.
- Added a PKCaptionsAssetResourceLoaderDelegate to support external captions.
Added a PKExternalSubtitle class to provide to the PKCaptionsAssetResourceLoaderDelegate.

### CODE REVIEW

**Comment by noamtamim** [comment]
AssetResourceLoaderDelegate refactor:
https://github.com/kaltura/playkit-ios/compare/develop...93eb3621de928f554e94e5bcfe41be5d45133a78

Captions feature:
https://github.com/kaltura/playkit-ios/compare/93eb3621de928f554e94e5bcfe41be5d45133a78...FEM-2357

File: Classes/Player/AVAsset/DefaultAssetHandler.swift [hunk 1]
```diff
@@ -14,16 +14,32 @@ import PlayKitUtils
...
+            asset.resourceLoader.setDelegate(assetLoaderDelegate, queue: resourceLoadingRequestQueue)
+        }
+    }
 
     required init() {
         
     }
     
+    private func replaceURL(_ url: URL, withScheme scheme: String) -> URL {
+        let urlString = url.absoluteString
```
Comment (noamtamim) [inline Classes/Player/AVAsset/DefaultAssetHandler.swift]: Consider using URLComponents for the replacement.

File: Classes/Player/AVAsset/DefaultAssetHandler.swift [hunk 2]
```diff
@@ -50,10 +66,22 @@ class DefaultAssetHandler: AssetHandler {
             return
         }
 
+        // Set the custom scheme for the external subtitles, if exists.
```
Comment (noamtamim) [inline Classes/Player/AVAsset/DefaultAssetHandler.swift:69]: Does it have to be in this class?

File: Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            }
+            
+            string.append("\n")
+            
+            allSubtitles.append(string)
+        }
+        return allSubtitles
+    }
+    
+    func getSubtitlem3u8(forSubtitle subtitle: PKExternalSubtitle) -> String {
```
Comment (noamtamim) [inline Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift]: Consider moving this function to PKExternalSubtitle (as `buildM3u8Playlist()` or similar).

File: Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        
+        let subtitlem3u8 = getSubtitlem3u8(forSubtitle: subtitle)
+        PKLog.debug("Subtitle (\(subtitleId)) m3u8:\n\(subtitlem3u8)")
+        let data = subtitlem3u8.data(using: .utf8)!
+        loadingRequest.dataRequest?.respond(with: data)
+        loadingRequest.finishLoading()
+        return true
+    }
+    
+    func getSubtitlesEXT() -> String {
```
Comment (noamtamim) [inline Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift]: Consider moving this function to PKExternalSubtitle (as `buildMasterLine()` or similar).

**Review by noamtamim — CHANGES_REQUESTED** [review]

File: Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift (hunk 3 above)
Comment (x-NR-x) [reply Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift]: Yes, you'r right, better approach. Will change it.

File: Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift (hunk 4 above)
Comment (x-NR-x) [reply Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift]: I will create an additional function in the PKExternalSubtitle as buildMasterLine().
But this function will still exist cause it goes over all the external subtitles.

File: Classes/Player/AVAsset/DefaultAssetHandler.swift (hunk 2 above)
Comment (x-NR-x) [reply Classes/Player/AVAsset/DefaultAssetHandler.swift:69]: Yes, cause I need to do it before creating the AVURLAsset.

File: Classes/Player/AVAsset/DefaultAssetHandler.swift (hunk 1 above)
Comment (x-NR-x) [reply Classes/Player/AVAsset/DefaultAssetHandler.swift]: Yes, will do. 
I copied it from the sample.

File: Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift [hunk 5]
```diff
@@ -16,7 +16,6 @@ class PKCaptionsAssetResourceLoaderDelegate: NSObject, AVAssetResourceLoaderDele
     
     private let extM3UPrefix = "#EXTM3U"
```
Comment (noamtamim) [inline Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift]: I don't think these constants should be instance fields. Either static fields or in the file itself, as `fileprivate` (preferred).

File: Classes/Player/Tracks/PKExternalSubtitle.swift [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+     
+        - autoSelect:       If the value is true, then the client MAY choose to play this Rendition in the absence of explicit user preference because it matches the current playback environment, such as chosen system language. This attribute is OPTIONAL.  Its absence indicates an implicit value of false.
+     
+        - forced:           A value of true indicates that the Rendition contains content that is considered essential to play. When selecting a FORCED Rendition, a client SHOULD choose the one that best matches the current playback environment (e.g., language). This attribute is OPTIONAL. It's absence indicates an implicit value of false.
+     
+        - language:         The value is a one of the standard Tags for Identifying Languages [RFC5646], which identifies the primary language used in the Rendition. This attribute is OPTIONAL.
+     
+        - characteristics:  The value contains one or more Uniform Type Identifiers [UTI] separated by comma (,) characters. This attribute is OPTIONAL. Each UTI indicates an individual characteristic of the Rendition.
+     
+        - vttURLString:     This is the string URL of the vtt.
```
Comment (noamtamim) [inline Classes/Player/Tracks/PKExternalSubtitle.swift]: Why is it optional? How does it work if not specified?

File: Classes/Player/Tracks/PKExternalSubtitle.swift [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+     
+        - isDefault:        If the value is true, then the client SHOULD play this Rendition of the content in the absence of information from the user indicating a different choice. This attribute is OPTIONAL. It's absence indicates an implicit value of false.
+     
+        - autoSelect:       If the value is true, then the client MAY choose to play this Rendition in the absence of explicit user preference because it matches the current playback environment, such as chosen system language. This attribute is OPTIONAL.  Its absence indicates an implicit value of false.
+     
+        - forced:           A value of true indicates that the Rendition contains content that is considered essential to play. When selecting a FORCED Rendition, a client SHOULD choose the one that best matches the current playback environment (e.g., language). This attribute is OPTIONAL. It's absence indicates an implicit value of false.
+     
+        - language:         The value is a one of the standard Tags for Identifying Languages [RFC5646], which identifies the primary language used in the Rendition. This attribute is OPTIONAL.
```
Comment (noamtamim) [inline Classes/Player/Tracks/PKExternalSubtitle.swift]: Should be a mandatory (but `Optional`) parameter. And it should be before the other parameters.

File: Classes/Player/Tracks/PKExternalSubtitle.swift [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        self.autoSelect = autoSelect
+        self.forced = forced
+        self.language = language ?? ""
+        self.characteristics = characteristics ?? ""
+        self.vttURLString = vttURLString
+        self.duration = duration
+    }
+    
+    public override var description: String {
+        return super.description + """
```
Comment (noamtamim) [inline Classes/Player/Tracks/PKExternalSubtitle.swift:59]: Maybe just return the master line as description?

**Review by noamtamim — CHANGES_REQUESTED** [review]

File: Classes/Player/Tracks/PKExternalSubtitle.swift (hunk 6 above)
Comment (x-NR-x) [reply Classes/Player/Tracks/PKExternalSubtitle.swift]: vttURLString is not optional.
characteristics is optional.

File: Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift (hunk 5 above)
Comment (x-NR-x) [reply Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift]: Created a struct with all the ones we are using.

File: Classes/Player/Tracks/PKExternalSubtitle.swift (hunk 8 above)
Comment (x-NR-x) [reply Classes/Player/Tracks/PKExternalSubtitle.swift:59]: I can, but I will also want to see the value of the given vttURLString and the given duration.

File: Classes/Player/Tracks/PKExternalSubtitle.swift (hunk 6 above)
Comment (noamtamim) [reply Classes/Player/Tracks/PKExternalSubtitle.swift]: It's the most important parameter, it should be moved to before the optional ones.
`init(id: String, name: String, language: String? = "", isDefault: Bool = false, autoSelect: Bool = false, forced: Bool = false, characteristics: String? = "", vttURLString: String, duration: Double)`

It's confusing that vttURLString and duration that don't have default values appear AFTER parameter that do have defaults.

**Review by noamtamim — APPROVED** [review]

**Pull request merged**